cpp/src/decisiontree/levelalgo/levelfunc_classifier.cuh

```diff
@@ -195,9 +195,12 @@ void grow_deep_tree_classification(
   sparsetree.resize(sparsetree.size() - lastsize);
   convert_scatter_to_gather(flagsptr, sample_cnt, n_nodes, nrows, d_nodecount,
                             d_nodestart, d_samplelist, tempmem);
+  if (tempmem->swap_depth == tree_params.max_depth - 1) {
```

Comment:
Why is this necessary?

Reply:
According to @vinaydes, the scatter loop (which only runs for depth > 14) exits one iteration before the `max_depth` (see line 202). So the handling of `depth_cnt` in the scatter loop is different from the gather loop. The edge case occurs where we set max_depth = 15, because the scatter loop does not run and `depth_cnt` never gets incremented. I found this fix by trial and error.

Maybe we should remove `depth_cnt` and write a separate function to compute depth by traversing the tree.

Reply:
I like this suggestion, as a tree traversal function is generally useful. That can be a separate change, though, no need to hold this one up.